File: mm_test.c

/* Some standard headers. */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* The timer, define CPU_TPS as the nr. of ticks per second on this CPU. */
#include "cycle.h"
#ifndef CPU_TPS
#define CPU_TPS 2.67e9
#endif


/* Main routine. */
int main ( int argc , char *argv[] ) {

    double *mat_a , *mat_b , *mat_c, irm = 1.0 / RAND_MAX;
    ticks start, finish;
    int i, j, k, n;
    
    /* Get n from the command line. */
    if ( argc < 2 ) {
        printf( "usage: %s <size of matrix>.\n" , argv[0] );
        return 0;
        }
    n = atoi( argv[1] );
    
    /* Allocate and fill the matrices with random numbers. */
    mat_a = (double *)malloc( sizeof(double) * n * n );
    mat_b = (double *)malloc( sizeof(double) * n * n );
    mat_c = (double *)malloc( sizeof(double) * n * n );
    for ( k = 0 ; k < n*n ; k++ ) {
        mat_a[k] = rand() * irm;
        mat_b[k] = rand() * irm;
        mat_c[k] = 0.0;
        }
        
    /* Tripple Do loop. */
    start = getticks();
    for ( i = 0 ; i < n ; i++ )
        for ( j = 0 ; j < n ; j++ )
            for ( k = 0 ; k < n ; k++ )
                mat_c[n*i+j] += mat_a[n*i+k] * mat_b[n*k+j];
    finish = getticks();
    printf( "triple do time %.15f\n" , (finish - start) / CPU_TPS );

    /* Free allocated memory. */
    free( mat_a ); free( mat_b ); free( mat_c );
    
    /* Goodbye. */
    return 0;

    }

File: mm_test.f90
program mm_test

  ! use mkl95_blas ! F95 wrappers for Intel's MKL
  implicit none

  real(8), allocatable :: mat_a(:,:), mat_b(:,:), mat_c(:,:)
  real(8)  :: start, finish
  integer  :: i, j, k, n, runs, r, status
  character*10 :: argv

  ! --- Get n from the command line
  call GET_COMMAND_ARGUMENT( 1 , argv , status=status )
  if ( status /= 0 ) then
    print*, 'usage: mm_test <nr of iterations> <size of matrix>.'
    call exit(0)
  endif
  read( argv , '(i10)' ) runs
  call GET_COMMAND_ARGUMENT( 2 , argv , status=status )
  if ( status /= 0 ) then
    print*, 'usage: mm_test <nr of iterations> <size of matrix>.'
    call exit(0)
  endif
  read( argv , '(i10)' ) n

  allocate (mat_a(n,n),mat_b(n,n))
  call random_number (mat_a)
  call random_number (mat_b)

  ! --- Fortrans 'matmul' intrinsic
  allocate (mat_c(n,n))
  call cpu_time(start)
  do r=1,runs
      mat_c=0.0d0
      mat_c=matmul(mat_a,mat_b)
  enddo
  call cpu_time(finish)
  print*, 'matmul    time',finish-start

  ! --- Triple Do loop
  call cpu_time(start)
  do r=1,runs
      mat_c=0.0d0
      do i=1,n
         do j=1,n
            do k=1,n
               mat_c(i,j)=mat_c(i,j)+mat_a(i,k)*mat_b(k,j)
            end do
         end do
      end do
  enddo
  call cpu_time(finish)
  print*, 'triple do time',finish-start

  deallocate (mat_a,mat_b,mat_c)

end program mm_test
